pkg/proto/address.go

```diff
@@ -17,179 +18,221 @@ import (
 )
 
 const (
-	headerSize     = 2
-	bodySize       = 20
-	checksumSize   = 4
-	AddressSize    = headerSize + bodySize + checksumSize
-	aliasFixedSize = 4
+	AddressIDSize            = 20
+	wavesAddressHeaderSize   = 2
+	wavesAddressBodySize     = AddressIDSize
+	wavesAddressChecksumSize = 4
+	WavesAddressSize         = wavesAddressHeaderSize + wavesAddressBodySize + wavesAddressChecksumSize
 
-	addressVersion byte = 0x01
-	aliasVersion   byte = 0x02
+	wavesAddressVersion byte = 0x01
+	aliasVersion        byte = 0x02
 
+	aliasFixedSize = 4
 	AliasMinLength = 4
 	AliasMaxLength = 30
 	AliasAlphabet  = "-.0123456789@_abcdefghijklmnopqrstuvwxyz"
 	AliasPrefix    = "alias"
 
+	EthereumAddressSize = AddressIDSize
+
 	MainNetScheme   byte = 'W'
 	TestNetScheme   byte = 'T'
 	StageNetScheme  byte = 'S'
 	CustomNetScheme byte = 'E'
 )
 
-// Address is the transformed Public Key with additional bytes of the version, a blockchain scheme and a checksum.
-type Address [AddressSize]byte
+type AddressID [AddressIDSize]byte
+
+func (a AddressID) Bytes() []byte {
+	return a[:]
+}
+
+type Address interface {
+	ID() AddressID
+	Bytes() []byte
+	String() string
+}
+
+// EthereumAddress is the first 20 bytes of Public Key's hash for the Waves address, or the 20 bytes of an Ethereum address.
+type EthereumAddress [EthereumAddressSize]byte
+
+// Bytes converts the fixed-length byte array of the EthereumAddress to a slice of bytes.
+func (b EthereumAddress) Bytes() []byte {
+	return b[:]
+}
+
+func (b EthereumAddress) ID() AddressID {
+	var id AddressID
+	copy(id[:], b[:])
+	return id
+}
+
+func (b EthereumAddress) String() string {
```

Comment:
Stringer implementation for EthereumAddress isn't correct. See correct implementation in go-ethereum repository. (But maybe for our purposes we don't need correct implementation of this method.)